### PR TITLE
[#162407499] Collect service metrics from log-cache

### PR DESCRIPTION
**Pull request by bandesz**
## What

The app will expose not just the application, but the service metrics - using Log Cache.

Log Cache doesn't have any Firehose endpoints (like Doppler), so we need to poll (by default every 60 seconds) a service specific endpoint for all metrics for the last 15 minutes and pick the newest ones.

We also refactored how the application metric collection worked, although the Watcher is mostly untouched.
Changes:
 - remove watcher manager
 - rename exporter to discovery
 - move all app related structs to the app package
 - move all cf client related functions into a custom client
 - use the real Prometheus registry when testing the app discovery
 - better error handling in watcher (previously firehose errors could cause some metrics to get stuck)
 - add better lifecycle management by using go contexts

We updated the README with more details about the metric collection.

## How to review

It is probably worth to run it with real services, register in a Prometheus server and check if the metrics look correct.

## Who can review it

Not @keymon or me.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -1,43 +1,69 @@
...
+	"github.com/alphagov/paas-prometheus-exporter/service"
+
 	"github.com/cloudfoundry-community/go-cfclient"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
 var (
 	version            = "0.0.3"
```
Comment (idavidmcdonald) [inline main.go]: Please bump this version as part of this PR :)

Comment (bandesz) [reply main.go]: 👍

**Comment by idavidmcdonald** [comment]
So I've had a quick glance at the code changes but my knowledge of golang probably isn't good enough to give this a proper code review. However I've managed to run this locally just to take a look and had a few questions/bits of feedback:

### Labels we may not need

There are several labels on the service metrics that exist that are always there and always the same value for example `ip`, `deployment`, `index`. Is this information that needs to be attributed to each timeseries? I think this is just contextual information and that if we needed it we could likely include it using an "info metric". You can see [an example of an info metric here](https://www.robustperception.io/exposing-the-software-version-to-prometheus) or if you find the Prometheus up and running book by RE Observe desks there is a better section about them. My list above of labels above may not be exhaustive (for example `source` and `origin` would also be contenders).

### Job label
Do we want to include a `job` label on our metrics? `job` is usually used for identifying a group of targets to be scraped and will be likely be renamed to `exported_job` due to relabelling as mentioned in https://prometheus.io/docs/concepts/jobs_instances/. This label may also come under the above comment about not being useful for our users or to go in an info metric.

### Metric naming
Good work on some metrics having their units appended e.g. `cpu_percent`.
Some metrics though have slightly stranger names such as `connections_conn` or `connection_errors_err` which could maybe be tidied up but this might be an issue with the source of your data rather than in the paas-prometheus-exporter itself.


I'm going to try and have a look at some of the tests around the app metrics later on today as I haven't yet had time.

**Comment by keymon** [comment]
> Labels we may not need
> Job label

I guess we could explicitly filter those out, they do not provide too much value. They come from the collector + loggregator stack.

> Metric naming

Yeah, that comes from the collector, and maybe the units do not make too much sense. https://github.com/alphagov/paas-rds-metric-collector/blob/master/pkg/collector/postgres_collector.go#L24

 `connections` and   `connection_errors` could be a `count`? IDK. Suggestions?

**Comment by idavidmcdonald** [comment]
> Yeah, that comes from the collector, and maybe the units do not make too much sense. https://github.com/alphagov/paas-rds-metric-collector/blob/master/pkg/collector/postgres_collector.go#L24
> 
> `connections` and `connection_errors` could be a `count`? IDK. Suggestions?

Regarding metric naming for connections, I guess my thoughts were that the unit `conn` and `err` was not useful on the end. I'd argue you don't need to have a unit on these. I've only glanced but potentially the paas-rds-metric-collector could have units as optional.

I'd also suggest maybe not using `count` as it might imply that it is a counter metric (one that can only rise in value or be reset to 0) whereas these metrics are gauges (can go up and down in value).

**Comment by keymon** [comment]
>  I'd argue you don't need to have a unit on these. I've only glanced but potentially the paas-rds-metric-collector could have units as optional.

Nah, the loggregator API requires you to put an unit in the gauges:  https://github.com/cloudfoundry/go-loggregator/blob/master/ingress_client.go#L253

We can add just so specific adhoc `if` to drop the unit in these cases.

**Comment by bandesz** [comment]
@idavidmcdonald I agree with your points.

I added two additional commits:
 - https://github.com/alphagov/paas-prometheus-exporter/pull/15/commits/20ec6c185c1d67b36ae8861b929ad14000a87681 - we'll exclude all internal labels (deployment, index, ip, job, origin), as these are not useful for tenants at all
 - https://github.com/alphagov/paas-prometheus-exporter/pull/15/commits/b4b42f248ba6ad6986625db2e2ce4c08655eab46 - we will only allow some whitelisted labels: percent, byte, bytes, s, second, seconds, ms

**Comment by venusbb** [comment]
PaaS team has done the code review on the PR and has addressed the comment - We believe Observe team can merge the PR if they are happy with the change.

**Comment by idavidmcdonald** [comment]
@bandesz this is looking good. Those commits are definite improvements. 

Small question - is there a particular reason why you didn't also remove the `source` label?

Also some of your units aren't plural such as `dbsize_byte` whereas Prometheus encourages you to use the plural (https://prometheus.io/docs/practices/naming/). Very happy to let this slide though so don't worry about it unless it's a 30 second fix.

**Comment by idavidmcdonald** [comment]
Cool, if there aren't any objections then I'll approve and merge this now. Thanks!

**Pull request merged**